types/sequencer.go

```diff
@@ -1,23 +1,162 @@
 package types
 
-import crypto "github.com/cosmos/cosmos-sdk/crypto/types"
+import (
+	"bytes"
+	"fmt"
 
-// TODO: remove this, as we can use the sequencers objects directly
+	pb "github.com/dymensionxyz/dymint/types/pb/dymint"
+	tmcrypto "github.com/tendermint/tendermint/crypto"
+	cmtproto "github.com/tendermint/tendermint/proto/tendermint/types"
+	"github.com/tendermint/tendermint/types"
+)
 
-// SequencerStatus defines the operating status of a sequencer
-type SequencerStatus int32
+// similar to tendermint/types/validator_set.go
+// but doesn't enforces proposer to be set, and keeps proposer hash
+type SequencerSet struct {
+	Sequencers   []*types.Validator `json:"sequencers"`
+	Proposer     *types.Validator   `json:"proposer"`
+	ProposerHash []byte
+}
 
-const (
-	// Proposer defines a sequencer that is currently the proposer
-	Proposer SequencerStatus = iota
-	// Inactive defines a sequencer that is currently inactive
-	Inactive
-)
+func (s *SequencerSet) GetProposerPubKey() tmcrypto.PubKey {
+	if s.Proposer == nil {
+		return nil
+	}
+	return s.Proposer.PubKey
+}
+
+// SetProposerByHash sets the proposer by hash.
+// It returns an error if the hash is not found in the sequencer set.
+// Used when updating proposer from the L2 blocks
+func (s *SequencerSet) SetProposerByHash(hash []byte) error {
+	for _, val := range s.Sequencers {
+		if bytes.Equal(GetHash(val), hash) {
+			s.SetProposer(val)
+			return nil
+		}
+	}
+	return fmt.Errorf("next sequencer not found in bonded set")
+}
+
+// SetProposer sets the proposer and adds it to the sequencer set.
+func (s *SequencerSet) SetProposer(proposer *types.Validator) {
+	s.Proposer = proposer
+	s.ProposerHash = GetHash(proposer)
+
+	// Add proposer to bonded set if not already present
+	if !s.HasAddress(proposer.Address) {
+		s.Sequencers = append(s.Sequencers, proposer)
+	}
```

Comment:
when can this occur?

Reply:
on edge case.
e.g where the proposer is updated from the hub, but for some reason we missed the events/sync of the sequencer list updates

Reply:
can you comment this